File: ftgo-kitchen-service/src/main/java/me/jangjunha/ftgo/kitchen_service/grpc/GrpcServer.java
package me.jangjunha.ftgo.kitchen_service.grpc;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.ServerInterceptors;
import io.grpc.health.v1.HealthCheckResponse;
import io.grpc.protobuf.services.HealthStatusManager;
import me.jangjunha.ftgo.common.auth.AuthInterceptor;
import me.jangjunha.ftgo.kitchen_service.service.KitchenService;
import me.jangjunha.ftgo.order_service.api.OrderServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.SmartLifecycle;

import java.io.IOException;
import java.util.Optional;

public class GrpcServer implements SmartLifecycle {

    private final Logger logger = LoggerFactory.getLogger(GrpcServer.class);
    private Optional<Server> server = Optional.empty();
    private final int port;
    private final HealthStatusManager health = new HealthStatusManager();

    private final KitchenService kitchenService;
    private final OrderServiceGrpc.OrderServiceBlockingStub orderService;


    public GrpcServer(int grpcServerPort, KitchenService kitchenService, OrderServiceGrpc.OrderServiceBlockingStub orderService) {
        this.port = grpcServerPort;
        this.kitchenService = kitchenService;
        this.orderService = orderService;
    }

    @Override
    public void start() {
        if (isRunning()) {
            return;
        }
        Server server = ServerBuilder.forPort(port)
                .addService(ServerInterceptors.intercept(
                        new KitchenServiceImpl(kitchenService, orderService),
                        new AuthInterceptor()
                ))
                .addService(health.getHealthService())
                .build();
        this.server = Optional.of(server);

        try {
            server.start();
        } catch (IOException e) {
            logger.error("gRPC server raises error", e);
            throw new RuntimeException(e);
        }
        health.setStatus("", HealthCheckResponse.ServingStatus.SERVING);
        startAwaitThread();
        logger.info("gRPC server started, listening on %d".formatted(port));
    }

    @Override
    public void stop() {
        server.ifPresent(server -> {
            logger.info("*** shutting down gRPC server since JVM is shutting down");
            server.shutdown();
            logger.info("*** gRPC server shut down");
        });
    }

    @Override
    public boolean isRunning() {
        return server.map(server -> !server.isTerminated()).orElse(false);
    }

    private void startAwaitThread() {
        server.ifPresent(server -> {
            Thread awaitThread = new Thread(() -> {
                try {
                    server.awaitTermination();
                } catch (InterruptedException e) {
                    logger.error("*** gRPC server awaiter interrupted");
                }
            });
            awaitThread.setName("grpc-server-awaiter");
            awaitThread.setDaemon(false);
            awaitThread.start();
        });
    }
}

File: ftgo-order-service/src/main/kotlin/me/jangjunha/ftgo/order_service/grpc/GrpcServer.kt
package me.jangjunha.ftgo.order_service.grpc

import io.grpc.Server
import io.grpc.ServerBuilder
import io.grpc.ServerInterceptors
import io.grpc.health.v1.HealthCheckResponse
import io.grpc.protobuf.services.HealthStatusManager
import me.jangjunha.ftgo.common.auth.AuthInterceptor
import me.jangjunha.ftgo.order_service.service.OrderService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.SmartLifecycle
import java.io.IOException

class GrpcServer(
    private val port: Int,
    private val orderService: OrderService,
) : SmartLifecycle {
    private val logger: Logger = LoggerFactory.getLogger(GrpcServer::class.java)

    private val health: HealthStatusManager = HealthStatusManager()
    private var server: Server? = null

    override fun start() {
        if (isRunning) {
            return
        }

        server = ServerBuilder.forPort(port)
            .addService(
                ServerInterceptors.intercept(
                    OrderServiceImpl(orderService),
                    AuthInterceptor()
                )
            )
            .addService(health.healthService)
            .build()
            .also { server ->
                try {
                    server.start()
                } catch (e: IOException) {
                    logger.error("gRPC server raises error", e)
                    throw RuntimeException(e)
                }
            }

        health.setStatus("", HealthCheckResponse.ServingStatus.SERVING)
        startAwaitThread()
        logger.info("gRPC server started, listening on %d".format(port))
    }

    override fun stop() {
        server?.let { server ->
            logger.info("*** shutting down gRPC server since JVM is shutting down")
            server.shutdown()
            try {
                server.awaitTermination()
            } catch (e: InterruptedException) {
                logger.error("*** gRPC server interrupted during termination", e)
            } finally {
                this.server = null
            }
            logger.info("*** gRPC server shut down")
        }
    }

    override fun isRunning(): Boolean {
        return server?.run { isShutdown } ?: false
    }

    private fun startAwaitThread() {
        val awaitThread = Thread {
            try {
                server?.awaitTermination()
            } catch (e: InterruptedException) {
                logger.error("*** gRPC server awaiter interrupted")
            }
        }
        awaitThread.name = "grpc-server-awaiter"
        awaitThread.isDaemon = false
        awaitThread.start()
    }
}
